students/AleGuardia/Lesson06/mailroom_four.py

```diff
@@ -0,0 +1,141 @@
+              f"{main_responses[1]}{main_responses[2]}{main_responses[3]}{main_responses[4]}"
+thank_you_note = "Dear {}:\n\nWe want to thank you for your generous donation of ${:.2f}.\n\n" \
+                 "It will be put to very good use.\n\n" \
+                 "\tSincerely,\n\t\tThe Team"
+
+
+def return_average(seq):
+    """Takes a sequence and returns the average"""
+    return sum(seq)/len(seq)
+
+
+def return_label():
+    """Builds and returns a table row label"""
+    label ="{:18} | {:>12} | {:>9} | {:>12}\n".format('Donor Name','Total Given','Num Gifts','Average Gift')
+    label += "-"*len(label)
+    return label
+
+
+def return_report(data):
+    """Creates and prints a report of donors and donations"""
+    report = f"{return_label()}\n"
+    for item in data:
+        row = "{:18} $ {:>12.2f}  {:>9d}  $ {:>12.2f}\n".format(item,sum(data[item]),len(data[item]),return_average(data[item]))
+        report += row
+    return report
+
+
+# Changed for loop to list Comprehension - added list comprehension
+def return_donors():
+    """Returns the names of all donors"""
+    # donors = ""
+    # for item in donations:
+    #     donors += item + "\n"
+    return [x + "\n" for x in donations]
+
+
+def prompt_donors():
+    """Prompt donor name and adds donor and donations"""
+    name = input("Please enter a donor name: ")
+    if name == 'quit':
+        return None
+    if name == 'list':
+        print(*return_donors())
+        return prompt_donors()
+    if name in donations:
+        return send_note(name,add_donation(name),thank_you_note)
+    if name not in donations:
+        add_donor(name)
+        return send_note(name,add_donation(name),thank_you_note)
+
+
+def add_donor(name):
+    """Adds a donor name"""
+    donations[name] = []
+    return donations
+
+
+# Added code to handle exceptions when user does not input a number
+def add_donation(name):
+    """Takes the name of a donor and adds a donation on his behalf"""
+    while True:
+        try:
+            amount = float(input("Please enter a donation amount for {}:".format(name)))
+            donations[name].append(float(amount))
+            return amount
+        except ValueError:
+            print("Please enter only digits")
+
+
+def send_note(name,donation,note):
+    """Takes a donor name,donation amount and a note and sends the note to the donor"""
+    print()
+    print(note.format(name,donation))
+    print()
+    return note.format(name,donation)
+
+
+def menu_selection(prompt, dispatch_dict, key_def=None):
+    while True:
+        response = input(prompt)
+        try:
+            if dispatch_dict[response]() == "quit":
+                break
+        except KeyError:
+            if not callable(key_def):
+                print("Please enter a valid number")
+            else:
+                key_def()
+
+
+def print_report():
+    print()
+    report = return_report(donations)
+    print(report)
+    return report
```

Comment:
your `print_report` function should only do print and not return a value; the unit test should test the `return_report` function onlyl.

note: a better name would be `get_report` (instead of `return_report`)